hy/compiler.py

```diff
@@ -1737,10 +1737,35 @@ def compile_require(self, expression):
         "unimport" it after we've completed `thing' so that we don't pollute
         other envs.
         """
-        expression.pop(0)
-        for entry in expression:
-            __import__(entry)  # Import it fo' them macros.
-            require(entry, self.module_name)
+        for entry in expression[1:]:
+            if isinstance(entry, HySymbol):
+                # e.g., (require foo)
+                __import__(entry)
+                require(entry, self.module_name, all_names=True,
+                        prefix=entry)
+            elif (isinstance(entry, HyList) and len(entry) == 2
```

Comment:
(1) Regarding repetition, I don't think there's enough of substance to factor out, because each branch is small but requires small differences; can you specify what kind of refactoring you'd like to see? (2) What should `(require [x [* *]])` do? Import a macro named `*` twice? `*` is a terrible name for a macro.